### DISCUSSION TITLE
1.4/2.0 way to analyze query ( previously `_join_entities`)

### DISCUSSION THREAD

**Discussion by jvanasco** [sqlalchemy/sqlalchemy#11398] [Usage Questions]
Like many others, I abused the private API and iterated over `Query._join_entities` to determine if a Class/table was involved in a query.  That no longer works.

Mike once pointed another user to #3225 as a future blue-sky way to do this.

That was abandoned in favor of #4472

I haven't been able to adapt anything from that approach, or the visitors module, to address my needs.

Does anyone have a suggestion on what I can use?

Context:

* TableA was split into TableA and TableA1
* Some filtering routines require TableA1 to be in the query, so the query needs to be inspected and the table joined onto it.

the pre-1.4 code was:

    _already_joined = any([i.entity == TableA1 for i in query._join_entities])
    if not _already_joined:
        query = query.join(TableA1)

**Comment by CaselIT** [comment]
Hi,

In select there is https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.Select.get_final_froms

Maybe you can use that?
